Add temporality awareness to openrag responses and chunk creation

EnjoyBacon7 opened the pull request:
API changes:

- OpenRAG now optionally expects "datetime" as as file metadata.
- Profiding "modified_at" and "created_at" in indexing requests overrides file-provided created and modified times
- "indexed_at" is ignored when provided in indexing request (it is overridden by the current date)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Temporal filtering for searches (created/modified/indexed/document dates) with automatic date extraction and configurable temporal weighting/decay
  * Temporal-aware reranking prioritizes recent results; search results surface temporal metadata and current-system timestamp in prompts

* **Documentation**
  * Expanded API and guides with temporal filtering examples and usage patterns

* **Chores**
  * Docker images updated to latest tags

* **Bug Fixes**
  * Timezone-aware handling of file timestamps for accurate metadata

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

EnjoyBacon7 committed:
Merge pull request #118 from linagora/dev

v1.1

EnjoyBacon7 committed:
Update Docker images to latest versions

EnjoyBacon7 committed:
Added temporal awareness to OpenRAG.

- openrag/utils/temporal.py
TemporalQueryNormalizer class extracts temporal filters
Date patterns recognition in multiple languages
Relative time Extraction

- openrag/components/indexer/chunker/chunker.py
Chunkers now add an indexed_at timestamp to documents
It is expected that indexed documents provide a created_at timestamp if available

- openrag/components/indexer/vectordb/vectordb.py
Milvus schema updated to include created_at and indexed_at fields
Added Temporal filtering support in vector database queries

- openrag/components/retriever.py & pipeline.py
Added temporal_filter parameter to all retrievers
Automatic temporal extraction from queries via TemporalQueryNormalizer
Injects current UTC datetime into system prompt

- openrag/components/reranker.py
Reranker now combines relevance and temporal scores using a linear decay formula

- RERANKER_TEMPORAL_WEIGHT (default 0.3)
- RERANKER_TEMPORAL_DECAY_DAYS (default 365)

EnjoyBacon7 committed:
Removed duplicate created_at field

EnjoyBacon7 committed:
modified_at and datetime metadata handling

Added extraction for "modified_at" field in indexation
Added "datetime" metadata field as preferred field for date information

EnjoyBacon7 committed:
Added chunk dates to prompt for LLM
Added formatted prompt logging in DEBUG mode
Fixed db search with date filters to use OR logic between date fields

paultranvan commented on prompts/example1/sys_prompt_tmpl.txt:
```diff
@@ -10,10 +12,17 @@ Prioritize **clarity, accuracy, and completeness** in your responses.
+4. Temporal Awareness
+   * Pay attention to the **temporal context** of both the query and the retrieved documents.
+   * Each document includes **creation_date** and **indexed_date** metadata indicating when it was created and indexed.
+   * When the user asks about **recent events**, **latest updates**, or uses temporal references (e.g., "last week", "yesterday", "this year"), prioritize documents with **more recent dates**.
```
Reformulation proposition with less redundancy: `When the query includes temporal references (e.g., "last week", "yesterday", "this year"), prioritize documents with **more recent dates**.`

EnjoyBacon7 committed:
Added documentation for temporal awareness API

paultranvan commented on openrag/utils/temporal.py:
```diff
@@ -0,0 +1,286 @@
+        self.relative_number_pattern = r'(\d+)\s*\w+|\w+\s+(\d+)'
+        
+        # English patterns for backward compatibility
+        self.english_patterns = {
```
The naming seems incorrect, it sounds more something like `common_languages_patterns`?

EnjoyBacon7 committed:
renamed var and updated temporal prompt

EnjoyBacon7 committed:
Updated temporal extraction

It is now a lot more conservative when interpreting dates, as such, it will less often interpret dates, but when it does, it is more accurate.

EnjoyBacon7 committed:
added suffix to date identification

EnjoyBacon7 committed:
added accent stripping

EnjoyBacon7 committed:
Fixing many edge cases and spelled out dates

EnjoyBacon7 committed:
fixed nullable blocking indexation

EnjoyBacon7 committed:
Insertions now initialize their fields

Since `nullable=True` does not seem to work as expected. All values are set to `None` if missing.

EnjoyBacon7 committed:
Fixed nullable issue v2

Strings cannot be null in Milvus. Since empty strings are falsy values in Python, its a drop-in fix.

EnjoyBacon7 commented:
This PR doesn't meet quality requirements and is superseded by #259
